src/dbnode/storage/limits/permits/types.go

```diff
@@ -43,15 +45,35 @@ type Manager interface {
 
 // Permits are the set of permits that individual codepaths will utilize.
 type Permits interface {
-	// Acquire blocks until an available resource is made available for the request permit
-	Acquire(ctx context.Context) error
+	// Acquire blocks until a Permit is available. The returned Permit is guaranteed to be non-nil if error is
+	// non-nil.
+	Acquire(ctx context.Context) (Permit, error)
 
 	// TryAcquire attempts to acquire an available resource without blocking, returning
-	// true if an resource was acquired.
-	TryAcquire(ctx context.Context) (bool, error)
+	// a non-nil a Permit if one is available. Returns nil if no Permit is currently available.
+	TryAcquire(ctx context.Context) (Permit, error)
 
-	// Release gives back one acquired permit from the specific permits instance. The user can pass an optional quota
-	// indicating how much of quota was used while holding the permit.
+	// Release gives back one acquired permit from the specific permits instance.
 	// Cannot release more permits than have been acquired.
-	Release(quota int64)
+	Release(permit Permit)
+}
+
+// Permit is granted to a caller which is allowed to consume some amount of quota.
+type Permit interface {
```

Comment:
hmm, callers should never invoke release or acquire, fair to say? I guess the panics help guard that?

Reply:
correct. I couldn't think of a better name. maybe `PrepareForRelease` and `PrepareForAcquire`?

Reply:
or maybe `PostRelease` and `PreAcquire`, kind of like hooks for the managers to call